### PR TITLE
fix: Wallet switch chain issue

### PR DESCRIPTION
**Pull request by memoyil**
---

## PR-Codex overview
This PR focuses on enhancing the handling of blockchain interactions, particularly for the Binance Smart Chain (BSC) and its integration with wallet functionalities. It introduces a new MEV guard chain configuration and improves error handling in RPC calls.

### Detailed summary
- Created `BSCMevGuardChain` in `mevGuardChains.ts` for BSC RPC configuration.
- Updated `useAccountLocalEventListener` and `useAccountEventListener` to use `useMemo` for `onDisconnect` handlers.
- Improved error handling in `fetchChunk.ts` for RPC errors using specific error codes.
- Enhanced `checkWalletSupportAddEthereumChain` to handle new error codes for wallet interactions.
- Refactored `useAddMevRpc` to utilize `addChain` for adding BSC MEV guard chain.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/web/src/views/Mev/hooks/index.ts [hunk 1]
```diff
@@ -12,20 +12,13 @@ async function checkWalletSupportAddEthereumChain(walletClient: WalletClient) {
...
-        },
-      ],
+      params: [{}],
     })
 
     console.error('lack of parameter, should be error')
     return false
   } catch (error) {
-    if ((error as any)?.code === -32602) {
+    if ([-32602, -32603].includes((error as any)?.code)) {
```
[the remark this answers is not in the document]
Comment (ChefJerry) [reply apps/web/src/views/Mev/hooks/index.ts:21]: kindly use `InvalidParamsRpcError.code` & `InternalRpcError.code`, import from viem, will more readable

File: apps/web/src/views/Mev/hooks/index.ts [hunk 2]
```diff
@@ -12,24 +12,17 @@ async function checkWalletSupportAddEthereumChain(walletClient: WalletClient) {
...
       // @ts-ignore
-      params: [
-        // mock data without key params nativeCurrency
-        {
-          chainId: '0x38', // Chain ID in hexadecimal (56 for Binance Smart Chain)
-          chainName: 'PancakeSwap MEV Guard',
-          rpcUrls: ['https://bscrpc.pancakeswap.finance'], // PancakeSwap MEV RPC}
-        },
-      ],
+      params: [{}],
```
Comment (ChefJerry) [inline apps/web/src/views/Mev/hooks/index.ts:15]: it is not follow the standard EIP https://eips.ethereum.org/EIPS/eip-3085, we cannot use the return value to guarantee if it support
image

File: apps/web/src/views/Mev/hooks/index.ts [hunk 3]
```diff
@@ -12,20 +12,13 @@ async function checkWalletSupportAddEthereumChain(walletClient: WalletClient) {
     await walletClient.request({
       method: 'wallet_addEthereumChain',
       // @ts-ignore
-      params: [
-        // mock data without key params nativeCurrency
-        {
-          chainId: '0x38', // Chain ID in hexadecimal (56 for Binance Smart Chain)
```
[the remark this answers is not in the document]
Comment (ChefJerry) [reply apps/web/src/views/Mev/hooks/index.ts:18]: it is not a good way to use add network methods to check if it support, or it should not be checked, just send method

File: packages/chains/package.json [hunk 4]
```diff
@@ -37,6 +37,9 @@
   "engines": {
     "node": ">=10"
   },
+  "dependencies": {
+    "viem": "catalog:"
```
[the remark this answers is not in the document]
Comment (chefjackson) [reply packages/chains/package.json:41]: pls keep chain pkg simple. DO not add extra dependencies even peer if possible

**Comment by Chef-Yogi** [comment]
@memoyil 
after took out chainId 
Safepal pops this black empty popup after we update the way 

CleanShot 2024-12-09 at 00 40 12@2x

**Comment by Chef-Yogi** [comment]
"it is not a good way to use add network methods to check if it supports, or it should not be checked, just send method"
@ChefJerry  I agree with what you say, but we need to check if wallet support to determine whether we should display the 
MEV toggle or not, if we always show the MEV toggle, some of the user will click the toggle and expect that will add MEV RPC to their wallet, but nothing gonna happen if the wallet doesn't support it.
The user will be confused in this  case

**Review by Chef-Yogi — DISMISSED** [review]

**Comment by Chef-Yogi** [comment]
CleanShot 2024-12-09 at 03 39 28@2x
Hi, @memoyil Sir
it seems safepal now can pass the test

**Comment by Chef-Yogi** [comment]
This is a special case, I’ve try that safepal can’t add a new rpc, although it can pass the test

**Comment by Chef-Yogi** [comment]
In this case may be we can add chainId for safepal

**Review by Chef-Yogi — APPROVED** [review]

**Pull request merged**